src/devices/deltadore_x3d.c

```diff
@@ -0,0 +1,191 @@
+}
+
+static int deltadore_x3d_decode(r_device *decoder, bitbuffer_t *bitbuffer)
+{
+    uint8_t const preamble[] = {
+            /*0xaa, 0xaa, */ 0xaa, 0xaa, // preamble
+            0x81, 0x69, 0x96, 0x7e       // sync word
+    };
+
+    data_t *data;
+
+    if (bitbuffer->num_rows != 1) {
+        return DECODE_ABORT_EARLY;
+    }
+
+    int row = 0;
+    // Validate message and reject it as fast as possible : check for preamble
+    unsigned start_pos = bitbuffer_search(bitbuffer, row, 0, preamble, sizeof (preamble) * 8);
+
+    if (start_pos == bitbuffer->bits_per_row[row]) {
+        return DECODE_ABORT_EARLY; // no preamble detected
+    }
+
+    // check min length
+    if (bitbuffer->bits_per_row[row] < 10 * 8) { //preamble(4) + sync(4) + len(1) + data(1)
+        return DECODE_ABORT_LENGTH;
+    }
+
+    uint8_t whitening_key_msb = 0x01;
+    uint8_t whitening_key_lsb = 0xff;
+
+    uint8_t len;
+    bitbuffer_extract_bytes(bitbuffer, row, start_pos + sizeof (preamble) * 8, &len, 8);
+
+    // dewhite length
+    ccitt_dewhitening(&whitening_key_msb, &whitening_key_lsb, &len, 1);
+
+    if (len > 64) {
+        if (decoder->verbose) {
+            fprintf(stderr, "%s: packet to large (%d bytes), drop it\n", __func__, len);
+        }
+        return DECODE_ABORT_LENGTH;
+    }
+
+    uint8_t frame[65] = {0};
+    frame[0] = len;
+
+    // Get frame (len include the length byte)
+    bitbuffer_extract_bytes(bitbuffer, row, start_pos + (sizeof (preamble) + 1) * 8, &frame[1], (len - 1) * 8);
+
+    // dewhite data
+    ccitt_dewhitening(&whitening_key_msb, &whitening_key_lsb, &frame[1], len - 1);
+
+    if (decoder->verbose > 1) {
+        bitrow_printf(frame, (len) * 8, "%s: frame data: ", __func__);
+    }
+
+    uint16_t crc = crc16(frame, len - 2, 0x1021, 0x0000);
+
+    if ((frame[len - 2] << 8 | frame[len - 1]) != crc) {
+        if (decoder->verbose) {
+            fprintf(stderr, "%s: CRC invalid %04x != %04x\n", __func__, frame[len - 2] << 8 | frame[len - 1], crc);
+        }
+        return DECODE_FAIL_MIC;
+    }
+
+    char frame_str[64 * 2 + 1] = {0};
+    // do not put lenght and crc to raw data
+    for (int i = 0; i < len - 3; ++i) {
+        sprintf(&frame_str[i * 2], "%02x", frame[i + 1]);
+    }
+
+    /* clang-format off */
+    data = data_make(
+            "model",        "",                 DATA_STRING, "DeltaDore X3D",
+            "raw",          "Raw data",         DATA_STRING, frame_str,
+            "mic",          "Integrity",        DATA_STRING, "CRC",
+            NULL);
+    /* clang-format on */
+    decoder_output_data(decoder, data);
+
+    return 1;
+}
+
+static char *output_fields[] = {
+        "model",
+        "raw",
+        "mic",
+        NULL,
+};
+
+r_device deltadore_x3d = {
+        .name        = "DeltaDore X3D devices",
+        .modulation  = FSK_PULSE_PCM,
+        .short_width = 25,
+        .long_width  = 25,
+        .reset_limit = 25600,
```

Comment:
this seems excessive, with whitening how many bit-width gaps do you expect worst-case?
E.g. 5 -> 25*5 µs, round up -> 160

Reply:
I don't know, I took this value from the analyzer.

Reply:
so 160 should enough?

Reply:
That was just a random example. How many consecutive 0's (before removing the whitening) are there in valid data?

Reply:
Hm, ok, in a very bad case, the whitened data gets all zero. 
The maximum packet length is 64. The first byte (length) is xored with 0xff, so it will never be zero, the second byte is always 0xff. 
Left 62. 
The chance that the CRC gets whitened to zero is also very low, so left 60.
The message id and retry count is always around position 15, so 45 left.
So my idea is 45 * 8 * 25 = 9000.
What do you think?

Reply:
That's a worst^2  case ;) Our detector will pick up the timing from the preamble, but even with that I guess that 5-10% margin/jitter is the maximum to work still. So no more than 10 to 20 zero bits will work reliably.
Maybe make it 9000, but it will very likely not work with that many zeros. (The reason Manchester code and similar are used).

Reply:
I'll make some tests.

Reply:
I drilled it down to 800, I think more than 4 bytes of zero should not happen.